src/textual/validation.py

```diff
@@ -0,0 +1,505 @@
+"""Framework for validating string values"""
+
+from __future__ import annotations
+
+import math
+import re
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field
+from typing import Callable, Sequence
+from urllib.parse import urlparse
+
+import rich.repr
+
+from textual._types import Pattern
+
+
+@dataclass
+class ValidationResult:
+    """The result of calling a `Validator.validate` method."""
+
+    failures: Sequence[Failure] = field(default_factory=list)
+    """A list of reasons why the value was invalid. Empty if valid=True"""
+
+    @staticmethod
+    def merge(results: Sequence["ValidationResult"]) -> "ValidationResult":
+        """Merge multiple ValidationResult objects into one.
+
+        Args:
+            results: List of ValidationResult objects to merge.
+
+        Returns:
+            Merged ValidationResult object.
+        """
+        is_valid = all(result.is_valid for result in results)
+        failures = [failure for result in results for failure in result.failures]
+        if is_valid:
+            return ValidationResult.success()
+        else:
+            return ValidationResult.failure(failures)
+
+    @staticmethod
+    def success() -> ValidationResult:
+        """Construct a successful ValidationResult.
+
+        Returns:
+            A successful ValidationResult.
+        """
+        return ValidationResult()
+
+    @staticmethod
+    def failure(failures: Sequence[Failure]) -> ValidationResult:
+        """Construct a failure ValidationResult.
+
+        Args:
+            failures: The failures.
+
+        Returns:
+            A failure ValidationResult.
+        """
+        return ValidationResult(failures)
+
+    @property
+    def failure_descriptions(self) -> list[str]:
+        """Utility for extracting failure descriptions as strings.
+
+        Useful if you don't care about the additional metadata included in the `Failure` objects.
+
+        Returns:
+            A list of the string descriptions explaining the failing validations.
+        """
+        return [
+            failure.description
+            for failure in self.failures
+            if failure.description is not None
+        ]
+
+    @property
+    def is_valid(self) -> bool:
+        return len(self.failures) == 0
+
+
+@dataclass
+class Failure:
+    """Information about a validation failure."""
+
+    validator: Validator
+    """The Validator which produced the failure."""
+    value: str | None = None
+    """The value which resulted in validation failing."""
+    description: str | None = None
+    """An optional override for describing this failure. Takes precedence over any messages set in the Validator."""
+
+    def __post_init__(self) -> None:
+        # If a failure message isn't supplied, try to get it from the Validator.
+        if self.description is None and self.validator is not None:
```

Comment:
`self.validator` can never be `None` according the the type.

I gather this sets `description` if its None. But that means that `description` is still type as being `| None`, which it never will be from the POV of the Textual developer.

Can this logic be moved to where a `Failure` is constructed? Thus removing the need to type `description` as potentially being `None` ?

Reply:
I'm not sure this is possible without having to duplicate this logic across all validators (meaning users writing custom validators would also have to duplicate this logic). I think things will become much messier and the trade-off isn't worth it unless you can suggest another approach. 

Is the description being `None` bad? Maybe as a developer who wants a custom `Validator`, I just have no need to describe my validation failure because I won't be using the description anyway.